### PR TITLE
fix(upstream): harden MCP proxy — process orphan, body OOM, UTF-8 panic

### PR DESCRIPTION
**Pull request by jmagar**
## Summary

Three independent defects in the upstream MCP proxy subsystem (`crates/lab/src/dispatch/upstream/`) surfaced during medium-effort code review. All three are exploitable in normal operation — F1 by any misbehaving stdio MCP server, F2 by a hostile or buggy upstream, F3 by any upstream that advertises a tool with a multi-byte unicode name when a wildcard exposure policy is set.

Plan: [`docs/superpowers/plans/2026-05-23-upstream-proxy-hardening.md`](docs/superpowers/plans/2026-05-23-upstream-proxy-hardening.md)
Beads: [`lab-4z8sx`](epic) → `.1` (F1), `.2` (F2), `.3` (F3)

## F1 — stdio child process-group orphan (HIGH)
**Commit:** `63239102` — RAII ProcessGroupGuard

`process_wrap::ProcessGroup::leader()` puts the spawned child in its own process group, but only the explicit async `UpstreamConnection::shutdown()` called `terminate_process_group_sigterm/sigkill`. When the connect future was dropped (15s `DISCOVERY_TIMEOUT` fires, `list_all_tools` errors, `buffer_unordered` cancellation), the Child's Drop only killed the lead PID — the actual MCP server (e.g. node forked by npx) was left attached to PID 1. Matches the documented `process_spawn_culprit` memory: lab spawning `codex-acp` via npx leaked the underlying node on every discovery hiccup.

**Fix:** new `ProcessGroupGuard` (`crates/lab/src/dispatch/upstream/process_guard.rs`) armed immediately after spawn, disarmed only on full success. `UpstreamConnection` gets its own `Drop` impl that performs the same TERM+KILL. `shutdown()` zeroes `runtime.pgid` BEFORE its first `.await` so the consuming Drop no-ops on the graceful path — no double-kill. Sync Drop fires TERM then KILL back-to-back; the kernel handles the race (Drop must not block).

Per research: do NOT add `Command.kill_on_drop(true)` — rmcp's `TokioChildProcess` already handles per-PID drop via its own async-kill path. Our guard targets the GROUP (killpg), which is additive to rmcp's per-PID drop.

## F2 — streaming response-body cap (HIGH)
**Commit:** `e15acec3` — BodyCappedHttpClient implementing rmcp's StreamableHttpClient

The 10 MB `DEFAULT_MAX_RESPONSE_BYTES` cap was being checked AFTER the upstream response was fully buffered and deserialized at `pool.rs:1748,2035,2532,2616`. A hostile upstream returning 5 GB OOMed the gateway before the check ran.

**Fix:** new `BodyCappedHttpClient` (`crates/lab/src/dispatch/upstream/http_client.rs`) wraps `reqwest::Client` and implements rmcp's `StreamableHttpClient` trait. Cap applied to **decoded** bytes BEFORE deserialization:

| Path | Cap semantics |
|------|---------------|
| `post_message` → `Json(_, _)` | Cumulative byte cap on the buffer (Content-Length pre-check + counting `bytes_stream`). |
| `post_message` → `Sse(_, _)` | **Per-event cap** — reject when a single SSE event exceeds N bytes. NOT cumulative. |
| `get_stream` | **Per-event cap** — same. |
| `delete_session` | No significant body; cap not applied. |

Per-event semantics for SSE matter — accumulating cumulative would disconnect any legitimate long-lived subscription after ~10 MB of cumulative bytes (engineering review caught this).

Cap applies to **decoded** bytes (reqwest auto-decodes gzip/br/zstd; `bytes_stream()` yields decoded chunks), so a 1 KB gzip-bomb expanding to 50 MB is rejected post-decompression.

**Known follow-up:** the OAuth HTTP path is NOT yet capped because `OauthClientCache` returns `Arc<AuthClient<reqwest::Client>>` and `AuthClient` is `#[non_exhaustive]`. Threading the cap requires the cache to build `AuthClient<BodyCappedHttpClient>` end-to-end. A WARN log documents the gap. Tracked as a follow-up in `lab-4z8sx.2`. Non-OAuth path (which is where hostile-upstream risk concentrates) is fully capped.

Defense-in-depth: existing post-hoc checks at `pool.rs:1748/2035/2532/2616` remain.

## F3 — UTF-8 char-boundary panic in wildcard_matches (MEDIUM)
**Commit:** `2e3f58eb` — `str::match_indices` rewrite + proptest

`candidate[cursor..].find(part)` in `types.rs:158` sliced at a byte offset that may not lie on a UTF-8 boundary after `cursor += part.len()` advanced past a matched ASCII byte sitting before a multi-byte char. Trigger: pattern `f*o` + candidate `f∂o` → `byte index 1 is not a char boundary` panic. Tool names come from upstream MCP servers (low trust).

**Fix:** rewrite using `str::match_indices` which always yields char-boundary-aligned byte offsets. Added regression test for the documented trigger, three unicode anchored cases, BIDI override (U+202E), and a `proptest` fuzz running 256+ random unicode strings against `*`-injected patterns.

Identical lesson previously learned at `lab-fstf.3` (mcpregistry pagination) — `str::is_char_boundary` / `match_indices` / `char_indices`.

## Pre-existing baseline fixes (necessary to land this branch)

- **`mcp/server.rs`** — added test-only `tool_search_schema_visible(auth)` wrapper for a pre-existing test that referenced a missing symbol. Without this, `cargo test --lib` doesn't compile on main.
- **`cli/gateway.rs`, `dispatch/gateway/dispatch.rs`, `mcp/server.rs`** — `cargo fmt` cleanup so `just lint` passes. These diffs were already on main; this branch just resolves them.

## Verification

```bash
just lint                                                # PASSES (clippy clean + fmt clean)
cargo test --workspace --all-features --lib              # 1366 passed, 26 ignored (1 pre-existing flake reproducible to wiremock port collision under parallel runs)
cargo test --workspace --all-features --lib -- dispatch::upstream
# 58 passed, 1334 filtered out
```

## Test plan

- [ ] CI green (lint + tests on Linux x86_64)
- [ ] Review F2 SSE per-event cap design (engineering-review BLOCKING finding now resolved)
- [ ] Confirm F1 Drop+shutdown invariant (`runtime.pgid` zero before await) under both graceful + abandoned paths
- [ ] Confirm F2 OAuth-path follow-up scope is acceptable

## Release Notes

* **Bug Fixes**
  * Fixed crashes when matching patterns with Unicode characters
  * Added safeguards to prevent out-of-memory conditions from oversized upstream responses
  * Improved graceful shutdown and cleanup for upstream connections
  * Enhanced HTTP header validation with proper error handling

* **Tests**
  * Expanded test coverage for Unicode edge cases and HTTP response scenarios
  * Added integration tests for response size limits

[Review Change Stack](https://app.coderabbit.ai/change-stack/jmagar/lab/pull/69?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: crates/lab/src/dispatch/upstream/http_client.rs [hunk 1]
```diff
@@ -0,0 +1,514 @@
...
+                if *count > max_u64 {
+                    let event_bytes = *count;
+                    *count = 0;
+                    Err(CappedStreamError::TooLarge {
+                        event_bytes,
+                        max_bytes,
+                    })
+                } else {
+                    if memmem(&chunk, b"\n\n") {
+                        *count = 0;
```
[the remark this answers is not in the document]
Comment (jmagar) [reply crates/lab/src/dispatch/upstream/http_client.rs]: Addressed in 3919c74a. The previous fix (chunk_contains_event_boundary with prev_lf state) handled the cross-chunk delimiter case but kept a 'add chunk.len() then reset' counter that could undercount a new event by up to ~1 chunk. The new account_event_bytes scans byte-by-byte for '\n' positions and resets the counter exactly at the '\n\n' delimiter, so bytes after the delimiter accumulate toward the next event correctly. Added 6 tests including a multi-event-per-chunk no-false-positive case.
